### PR TITLE
Deterministic seeds in debug mode

### PR DESCRIPTION
**Pull request by ksztenderski**
Signed-off-by: ksztenderski <ksztenderski@nvidia.com>

## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR

Add support for deterministic seed generation in debug mode when using pipeline seed.
Currently seed generation differ between standard and debug mode, in the future we might standardize this.

#### Additional information
- Affected modules and functionalities:

- Key points relevant for the review:

### Tests
- [x] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by ksztenderski** [comment]
!build

File: dali/test/python/test_pipeline_debug.py [hunk 1]
```diff
@@ -315,3 +315,38 @@ def test_init_config_pipeline():
...
+    pipe1 = seed_pipeline()
+    pipe2 = seed_pipeline()
+    compare_pipelines(pipe1, pipe2, 8, 10)
+
+
+@pipeline_def(batch_size=8, num_threads=3, device_id=0, seed=47, debug=True)
+def seed_rn50_pipeline_base():
+    rng = fn.random.coin_flip(probability=0.5)
+    jpegs, labels = fn.readers.file(
+        file_root=file_root, shard_id=0, num_shards=2)
```
Comment (klecki) [inline dali/test/python/test_pipeline_debug.py:339]: Maybe add random shuffle to the input data, that would be a good test for seed propagation.
```suggestion
        file_root=file_root, shard_id=0, num_shards=2, random_shuffle=True)
```

File: dali/python/nvidia/dali/_debug_mode.py [hunk 2]
```diff
@@ -131,6 +131,11 @@ def __init__(self, exec_func, **kwargs):
         self._cur_subpipeline_id = -1
         self._exec_func = exec_func
 
+        import numpy as np
+        if 'seed' in kwargs:
+            np.random.seed(kwargs['seed'])
+        self._seeds = np.random.randint(-2**63, 2**63-1, 1024)
```
Comment (klecki) [inline dali/python/nvidia/dali/_debug_mode.py:137]: Doesn't this reseed the global state? I think such usage is discouraged. 
As an alternative we can create a new random generator, seeding it with user provided seed: https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.default_rng

In that case we probably don't need to keep the list of initial seeds (up to you), we can run our private generator on demand.
Also, I don't know how DALI works when you pass negative seeds to it, but I think -1 might have a special meaning. I would suggest checking that or sticking to non-negative or positive range.

**Review by klecki — COMMENTED** [review]
Please use private random generator instead of relying on global state.

File: dali/test/python/test_pipeline_debug.py (hunk 1 above)
Comment (ksztenderski) [reply dali/test/python/test_pipeline_debug.py]: done

File: dali/python/nvidia/dali/_debug_mode.py (hunk 2 above)
Comment (ksztenderski) [reply dali/python/nvidia/dali/_debug_mode.py]: done

Comment (ksztenderski) [reply dali/python/nvidia/dali/_debug_mode.py]: I tried to imitate seed generation from backend, which uses fixed size list of random values that can also be negative.

Comment (ksztenderski) [reply dali/python/nvidia/dali/_debug_mode.py]: done

**Comment by ksztenderski** [comment]
!build

**Review by stiepan — APPROVED** [review]

File: dali/python/nvidia/dali/_debug_mode.py [hunk 3]
```diff
@@ -131,6 +131,10 @@ def __init__(self, exec_func, **kwargs):
         self._cur_subpipeline_id = -1
         self._exec_func = exec_func
 
+        import numpy as np
+        seed = kwargs.get('seed', np.random.randint(0, 2**32))
```
Comment (klecki) [inline dali/python/nvidia/dali/_debug_mode.py:135]: Just a nitpick: how it would work if someone passes `-1` (as it's the default value that indicates random seed in regular pipeline).

I'm wondering if we maybe should repeat the regular pipeline constructor keyword args in this constructor as well - but that would be a separate PR probably. Let's discuss.

**Comment by ksztenderski** [comment]
!build

File: dali/python/nvidia/dali/_debug_mode.py (hunk 3 above)
Comment (ksztenderski) [reply dali/python/nvidia/dali/_debug_mode.py]: done, generating new seed when seed < 0 (same as in regular Pipeline)

**Review by klecki — APPROVED** [review]
